### PR TITLE
Backport compiler determinism to 2.12.x [ci: last-only]

### PR DESCRIPTION
**Pull request by retronym**
Backports:

 - #6279
 - #6285
 - #6286
 - #6300
 - #6301
 - #6302
 - #6923

Which together resolve scala/scala-dev#405

A risk of the backport is that we create binary incompatibilities
with code compiled with `-opt:inline` that refers to fresh names
that are now changed. I had this problem with partest on 2.13.x:

https://github.com/scala/sbt-scala-module/commit/7aee2b537d30f227f2bb27ac6b1c09d5e94dcdf1#diff-b296be07ef3dd86a176475283d8583a4

Another risk is that there are serialization incompatibilities
that aren't covered by `SerializationStabilityTest`.

### CODE REVIEW

**Comment by adriaanm** [comment]
Going through the reviews of the 2.13 versions. Some concerns (will edit as I make my way through):

  - should we avoid breaking macro annotations? https://github.com/scala/scala/pull/6286/commits/87be453c234b53ef4550ce6eb932952f09b4bf7a

**Comment by adriaanm** [comment]
Preparing macro-annot PR at https://github.com/adriaanm/paradise/pull/new/2.12.8

**Comment by mkeskells** [comment]
Great to see this being backported. 
This will open up new possibilities with our CI plant both for reductions in compile and for test

**Comment by adriaanm** [comment]
I added a community build for 2.12.8-bin-e764204-SNAPSHOT to the queue (https://scala-ci.typesafe.com/job/scala-2.12.x-integrate-community-build/3650). It's using a custom config (https://github.com/adriaanm/community-builds/commit/bddc889cf80416eb905ac4db39752b89632c340b) that uses macro-paradise at https://github.com/scalamacros/paradise/tree/2.12.8 (I incorporated the community build fixes)

/cc @SethTisue FYI on queuing the build

**Review by adriaanm — APPROVED** [review]

**Comment by adriaanm** [comment]
After a green community build, this is ready to merge!

**Comment by SethTisue** [comment]
there were some failures in run 3650. the enumeratum one, at least, looks spurious, but let's see how a rebuild does: 
https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3652/

**Comment by adriaanm** [comment]
```
unexpected FAILED: classutil
unexpected FAILED: curryhoward
unexpected FAILED: enumeratum
unexpected FAILED: lift-json
unexpected FAILED: scala-refactoring
unexpected FAILED: scalatex
SUCCESS 179 FAILED 6 DID NOT RUN 4 TOTAL 189
UNEXPECTED 6
BLOCKERS:
2 enumeratum
1 lift-json
1 classutil
```

**Comment by adriaanm** [comment]
All test failures -- could they all be spurious?? Hitting rebuild one more time

**Comment by SethTisue** [comment]
for comparison, we should see whether any of the same failures crop up in https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3653/ which is a plain run that's running now

**Comment by retronym** [comment]
The lift-json looks like a change in behaviour, or bug, in scalap. Scalap contains an independant implementation of the unpickler.

I thought the changes to the pickler were backwards compatible, I just changed made the order of entries predictable.

```
[lift-json] [info] 
[lift-json] [error] ! Either can't be deserialized with type hints
[lift-json] [error]  net.liftweb.json.MappingException: Can't find constructor for class net.liftweb.json.Eith (Meta.scala:227)
[lift-json] [error] net.liftweb.json.Meta$.fail(Meta.scala:227)
[lift-json] [error] net.liftweb.json.ScalaSigReader$.$anonfun$readConstructor$1(ScalaSig.scala:25)
[lift-json] [error] net.liftweb.json.ScalaSigReader$.readConstructor(ScalaSig.scala:25)
[lift-json] [error] net.liftweb.json.Meta$.$anonfun$mappingOf$5(Meta.scala:143)
[lift-json] [error] net.liftweb.json.Meta$.parameterizedTypeOpt$1(Meta.scala:141)
[lift-json] [error] net.liftweb.json.Meta$.mkConstructor$1(Meta.scala:152)
```

Oh, that's not scalap, it is yet-another-unpickler implementation.

**Comment by retronym** [comment]
I've pushed a fix for lift-json.

Scheduled a community build: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3655/

**Comment by adriaanm** [comment]
I merged Seth's changes to the community build config into my branch that points to a different branch of paradise, and bumped the scala version to use 3dcbf47. Let's see how this one fares: https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3656

**Comment by SethTisue** [comment]
followup run https://scala-ci.typesafe.com/view/scala-2.12.x/job/scala-2.12.x-integrate-community-build/3662/

failed: classutil, curryhoward, scala-refactoring, enumeratum, scalatex

the enumeratum failure is unrelated (and now fixed by scala/community-builds/commit/32151444479ae4e1ecdc8d0202336ff64913f0a6)

the scalatex failure is unrelated

classutil, curryhoward, scala-refactoring look like they might be overly sensitive tests, hard to tell at a glance

only 3 projects didn't run because of upstream failures (pureconfig, refined, scalasti)

**Comment by retronym** [comment]
`classutil` is a picky test. Submitted a patch in https://github.com/bmc/classutil/pull/31

**Comment by retronym** [comment]
The `curryhoward` also appears to be benign. Making its test more tolerant: https://github.com/Chymyst/curryhoward/pull/32

**Comment by retronym** [comment]
Scala refactoring: https://github.com/scala-ide/scala-refactoring/pull/201

**Comment by retronym** [comment]
I'm happy that the community build hasn't found any bugs in this change. @SethTisue Should we update the community build to use my patches to the three projects above before we merge this?

**Comment by SethTisue** [comment]
I think due diligence has now been done, I don't think it's necessary to do another run

**Comment by mkeskells** [comment]
Is this waiting on any more reviews or can it be merged ?
We are keen to trial this in our CI toolchain

File: src/scalap/scala/tools/scalap/scalax/rules/scalasig/Symbol.scala
Comment (lrytz) [inline src/scalap/scala/tools/scalap/scalax/rules/scalasig/Symbol.scala:30]: This is mabye a little ad-hoc, to filter class type params, but not method type params?

**Comment by lrytz** [comment]
It looks ready, but I'll let @adriaanm merge it.

**Pull request merged**
